### PR TITLE
[WIP] Agent provisioning refactor

### PR DESCRIPTION
**Pull request by tmarkovski**
- Adds the ability to supply custom `ProvisioningRecord` with an invocation method to run post-provisioning configuration (e.g. register NYM for issuer/agent, etc).
- Simplify the agent service registration and building pipeline
  - Moved the agent provisioning to a hosted service `IHostedService`. This ensures the provisioning will be done during the service startup and also provides tighter integration with the Generic Host improvements coming in NETCore 3.0
  - Removed a lot of the redundant agent builder classes during configuration
  - Added agent factory support that allows creating agent instances with a default implementation of `IAgentFactory` [WIP]

### CODE REVIEW

File: src/AgentFramework.AspNetCore/MessageResponse.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
```
Comment (tplooker) [inline src/AgentFramework.AspNetCore/MessageResponse.cs]: Perhaps message context should also have a `BinaryWriter`, so we have symmetry in how we handle both request and response?

File: src/AgentFramework.AspNetCore/Middleware/AgentMiddleware.cs [hunk 2]
```diff
@@ -14,32 +15,31 @@ namespace AgentFramework.AspNetCore.Middleware
...
+            _contextProvider = contextProvider;
         }
 
-        /// <summary>Called by the ASPNET Core runtime</summary>
-        /// <param name="context">The context.</param>
-        /// <returns></returns>
-        /// <exception cref="Exception">Empty content length</exception>
-        public async Task InvokeAsync(HttpContext context)
+        /// <inheritdoc />
+        public async Task InvokeAsync(HttpContext context, RequestDelegate next)
```
Comment (tplooker) [inline src/AgentFramework.AspNetCore/Middleware/AgentMiddleware.cs]: So the way this middleware is now registered is that a post to any endpoint with mime type AgentWireMessage will invoke this?

File: src/AgentFramework.AspNetCore/DefaultAgent.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+using AgentFramework.Core.Contracts;
+using AgentFramework.Core.Handlers;
+using AgentFramework.Core.Messages;
+
+namespace AgentFramework.AspNetCore
+{
+    /// <summary>
+    /// Default agent.
+    /// </summary>
+    public class DefaultAgent : AgentMessageProcessorBase, IAgent
```
Comment (tplooker) [inline src/AgentFramework.AspNetCore/DefaultAgent.cs]: With this paradigm should AgentMessageProcessorBase be renamed to AgentBase?

File: src/AgentFramework.AspNetCore/DefaultAgent.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        /// <summary>
+        /// Gets the handlers.
+        /// </summary>
+        /// <value>The handlers.</value>
+        public IList<IMessageHandler> Handlers => _handlers;
+
+        /// <summary>
+        /// Configures the handlers.
+        /// </summary>
+        protected override void ConfigureHandlers()
```
Comment (tplooker) [inline src/AgentFramework.AspNetCore/DefaultAgent.cs]: I feel like this is an opportunity to review how we are managing these handlers, as `AgentMessageProcessorBase` is going to continue to bloat potentially. With the factory pattern you have established around building this default agent, perhaps the solution is to inject the handlers configuration via the factory? Or the handler configuration comes included with the AgentContext.

Comment (tplooker) [reply src/AgentFramework.AspNetCore/DefaultAgent.cs]: Perhaps the `SupportedMessages` currently on the IAgentContext should change to represent the protocols the agent supports and that configuration should be populated when the context is built.

**Comment by tplooker** [comment]
Following our conversation perhaps review the role of the AgentFactory and its distinction in regards to AgentContextProvider.

File: src/AgentFramework.AspNetCore/DefaultAgent.cs (hunk 4 above)
Comment (tmarkovski) [reply src/AgentFramework.AspNetCore/DefaultAgent.cs]: Added.

**Comment by tmarkovski** [comment]
> Following our conversation perhaps review the role of the AgentFactory and its distinction in regards to AgentContextProvider.

I merged them together into one provider. Let me know your thoughts.

File: src/AgentFramework.Core.Handlers/Agents/AgentBase.cs [hunk 5]
```diff
@@ -177,7 +187,7 @@ private async Task<MessageContext> UnpackAsync(IAgentContext agentContext, Messa
             return message;
         }
 
-        private IList<MessageType> GetSupportedMessageTypes() => _handlers.SelectMany(_ => _.SupportedMessageTypes, (parent, child) => new MessageType(child)).ToList();
+        private IList<MessageType> GetSupportedMessageTypes() => _handlers.SelectMany(x => x.SupportedMessageTypes).ToList();
```
Comment (tplooker) [inline src/AgentFramework.Core.Handlers/Agents/AgentBase.cs]: I think this should be public and use in the AgentProvider for populating the Agent contexts supported message types?

File: src/AgentFramework.Core.Handlers/Agents/AgentBase.cs [hunk 6]
```diff
@@ -86,13 +93,13 @@ protected AgentMessageProcessorBase(IServiceProvider provider)
         /// <exception cref="Exception">Expected inner message to be of type 'ForwardMessage'</exception>
         /// <exception cref="AgentFrameworkException">Couldn't locate a message handler for type {messageType}</exception>
         /// TODO should recieve a message context and return a message context.
-        protected async Task<byte[]> ProcessAsync(IAgentContext context, MessageContext messageContext)
+        public async Task<MessageResponse> ProcessAsync(IAgentContext context, MessageContext messageContext)
         {
             EnsureConfigured();
 
             context.SupportedMessages = GetSupportedMessageTypes();
```
Comment (tplooker) [inline src/AgentFramework.Core.Handlers/Agents/AgentBase.cs]: This is already populated in AgentContext so we can remove this? Also we need logic here to check that the AgentContext does not contain supported message types that are not available in the environment?

Comment (tmarkovski) [reply src/AgentFramework.Core.Handlers/Agents/AgentBase.cs]: I think the agent should throw if a handler is added that's not in the environment i.e. DI returns null.

File: src/AgentFramework.Core.Handlers/Agents/AgentBase.cs (hunk 5 above)
Comment (tmarkovski) [reply src/AgentFramework.Core.Handlers/Agents/AgentBase.cs]: Good catch. Added.

File: src/AgentFramework.Core.Handlers/Agents/AgentBase.cs (hunk 6 above)
Comment (tmarkovski) [reply src/AgentFramework.Core.Handlers/Agents/AgentBase.cs]: Updated.

File: src/AgentFramework.AspNetCore/MessageResponse.cs (hunk 1 above)
Comment (tmarkovski) [reply src/AgentFramework.AspNetCore/MessageResponse.cs]: I don't have any great ideas how to do this to achieve good usability without adding it for the sake of adding it.

Comment (tplooker) [reply src/AgentFramework.AspNetCore/MessageResponse.cs]: Fair enough we can differ until later

File: src/AgentFramework.Core.Handlers/Agents/DefaultAgentProvider.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        /// </summary>
+        /// <param name="walletOptions">Wallet options.</param>
+        /// <param name="poolOptions">Pool options.</param>
+        /// <param name="serviceProvider">Service provider.</param>
+        /// <param name="walletService">Wallet service.</param>
+        /// <param name="poolService">Pool service.</param>
+        public DefaultAgentProvider(
+            IOptions<WalletOptions> walletOptions,
+            IOptions<PoolOptions> poolOptions,
+            IAgent serviceProvider,
```
Comment (tplooker) [inline src/AgentFramework.Core.Handlers/Agents/DefaultAgentProvider.cs]: This should be defaultAgent?

**Pull request merged**